File: dynamic_listings/python/default/design_patterns/chain_of_responsibility.py
from typing import Self


class Handler:
    # This is the handler abstract/class interface that the sender connects to
    next_handler: Self = None  # The next handler in the chain

    def handle_request(self) -> None:
        if condition:
            # In case I can handle this request
            return self.real_handler()

        if self.next_handler:
            return self.next_handler.handle_request()

    def real_handler(self) -> None:
        # This function gets implemented in the concrete classes
        raise NotImplementedError("You should implement this!")

    def add_handler(self, new_handler: Self) -> Self:
        self.next_handler = new_handler
        return self.next_handler  # Allows for chaining .add_handler().add_handler()...

File: dynamic_listings/rust/default/design_patterns/chain_of_responsibility.rs
# #![allow(dead_code)]
trait Handler {
    fn handle_request(&mut self) {
#       let condition = true;
        if condition {
            return self.real_handler();
        }

        if let Some(next) = self.next() {
            return next.handle_request();
        }
    }

    fn real_handler(&mut self);
    fn next(&mut self) -> &mut Option<Box<dyn Handler>>;
    fn add_handler(&mut self, new_handler: &dyn Handler);
}
